### DISCUSSION TITLE
Optional named arguments

### DISCUSSION THREAD

**Discussion by boan-jfm** [libvips/php-vips#220] [Q&A] [answered]
We have cropped a PNG image using PHP-VIPS and ImageMagick. PHP-VIPS creates the smallest file-size (~7KB) vs ImageMagick (~24KB) with standard settings. However, ImageMagick's version does have more details. How do we configure the options?

We can see that the crop method takes some optional named arguments:

https://www.libvips.org/API/current/libvips-conversion.html#vips-crop
![image](https://github.com/libvips/php-vips/assets/13201214/e86e6317-dfaf-4c92-8f2d-ecef78714330)

https://github.com/libvips/php-vips/blob/v2.2.0/src/ImageAutodoc.php#L170

But where can we find the arguments?

Thanks

P.S. It would be nice to have a search option here: https://www.libvips.org/API/current/

**Comment by jcupitt** [comment, marked as answer]
Hi @boan-jfm,

You need `pngsave`:

https://www.libvips.org/API/current/VipsForeignSave.html#vips-pngsave

You can find these things by ^F on the function list:

https://www.libvips.org/API/current/func-list.html

which lists operations by what they do, or by ^F on the API index:

https://www.libvips.org/API/current/api-index-full.html

which indexes by the C function name.

A better search would be nice, I agree :(

It's unfortunate that phpdoc does not support optional arguments (as far as I know).

**Reply by boan-jfm** [reply]
Thank you, it works.

We used [writeToFile](https://github.com/libvips/php-vips/blob/v2.2.0/src/Image.php#L955), which also takes some options, but they a not documented here: https://www.libvips.org/API/current/VipsImage.html#vips-image-write-to-file

Autocompletion using copilot showed the option `['Q' => 100]`, which seems to effect the quality as well  - where can we find the arguments if we need to know this in the future?

Thanks :-)

**Reply by jcupitt** [reply]
`writeToFile` just calls one of the savers for you, so you can pass any options which the selected saver takes. For example:

```php
$image->writeToFile("x.png", ["palette" => true]);
```

The `vips` CLI is handy for seeing arguments, eg.:

```
john@banana ~ $ vips pngsave
save image to png file
usage:
   pngsave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   compression  - Compression factor, input gint
			default: 6
			min: 0, max: 9
   interlace    - Interlace image, input gboolean
			default: false
   profile      - ICC profile to embed, input gchararray
   filter       - libpng row filter flag(s), input VipsForeignPngFilter
   palette      - Quantise to 8bpp palette, input gboolean
			default: false
   Q            - Quantisation quality, input gint
			default: 100
			min: 0, max: 100
   dither       - Amount of dithering, input gdouble
			default: 1
			min: 0, max: 1
   bitdepth     - Write as a 1, 2, 4, 8 or 16 bit image, input gint
			default: 8
			min: 1, max: 16
   effort       - Quantisation CPU effort, input gint
			default: 7
			min: 1, max: 10
   strip        - Strip all metadata from image, input gboolean
			default: false
   background   - Background value, input VipsArrayDouble
   page-height  - Set page height for multipage save, input gint
			default: 0
			min: 0, max: 10000000
operation flags: sequential nocache 
john@banana ~ $ 
```

**Reply by jcupitt** [reply]
(copilot will know very few of the optional arguments, and often get them wrong, unfortunately)
